smack-tcp/src/main/java/org/jivesoftware/smack/tcp/XmppTcpTransportModule.java

```diff
@@ -580,6 +582,24 @@ protected XmppTcpNioTransport(ModularXmppClientToServerConnectionInternal connec
             super(connectionInternal);
         }
 
+        @Override
+        public StreamOpenFactory getStreamOpenFactory() {
+            return new StreamOpenFactory() {
+                @Override
+                public StreamOpen getStreamOpen() {
+                    CharSequence to = connectionInternal.connection.getXMPPServiceDomain();
+                    CharSequence from = null;
+                    CharSequence localpart = connectionInternal.connection.getConfiguration().getUsername();
+                    if (localpart != null) {
+                        from = XmppStringUtils.completeJidFrom(localpart, to);
+                    }
+                    String id = connectionInternal.connection.getStreamId();
+                    StreamOpen streamOpen = new StreamOpen(to, from, id, connectionInternal.connection.getConfiguration().getXmlLang(), StreamOpen.StreamContentNamespace.client);
```

Comment:
Now this duplicates code from `AbstractXMPPConnection.sendStreamOpen()`, but after thinking a little bit about it, I can't come up with a nice way to deduplicate this. I'd say we keep it for now as is, just wanted to point that out.

Reply:
Thanks for pointing that out. However, I have tried removing `AbstractXMPPConnection.sendStreamOpen()` and found that it still is required for non-modular connections. So lets keep it until `XmppTCPConnection/ XMPPBOSHConnection` is deprecated :)